sqlsmith: add support for DO block

ZhouXing19 opened the pull request:
informs https://github.com/cockroachdb/cockroach/issues/106368

A `DO` block can be either a top-level sql statement or be part of the plpgsql function body. In this commit we introduce the DO block for both of these 2 usages. We also introduced `DisableDoBlocks` which disables both usages of a DO block.

Release note: None

yuzefovich reviewed (commented):
> Reviewed 3 of 3 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/151983)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @DrewKimball)

___
*[`-- commits` line 2 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/151983#-OXyV2Sm5jHSqr8y5hwf:-OXyV2Sm5jHSqr8y5hwg:bff6fq7):*
nit: we should update the commit message.
___
*[`pkg/internal/sqlsmith/plpgsql.go` line 144 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/151983#-OXyVm3Q6cs9vngiks6I:-OXyVm3Q6cs9vngiks6J:b-qvetml) ([raw file](https://github.com/cockroachdb/cockroach/blob/7be608d016075aa34516ed500046491d605ee922/pkg/internal/sqlsmith/plpgsql.go#L144)):*
> ```Go
> 		{2, makePLpgSQLIf},
> 		{2, makePLpgSQLWhile},
> 		{2, makeDoBlockWithinScope},
> ```

Given that DO block is only allowed as top-level SQL stmt, I don't think we should be including it into `plpgsqlStmts`. In other words, we should only be generating the DO block at the level of `makeInsert` or `makeSelect`.

We definitely want to somehow include it into `allStatements` in `relational.go`. Perhaps we need to have two variants like `makeDoBlockNonMutating` (which cannot execute any mutations) and general `makeDoBlock` (that can execute any stmts within).
___
*[`pkg/internal/sqlsmith/relational.go` line 921 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/151983#-OXyWlvMBAg9uAFN70b0:-OXyWlvMBAg9uAFN70b1:bfm0hir) ([raw file](https://github.com/cockroachdb/cockroach/blob/7be608d016075aa34516ed500046491d605ee922/pkg/internal/sqlsmith/relational.go#L921)):*
> ```Go
> 
> func makeDoBlock(s *Smither) (tree.Statement, bool) {
> 	if s.disableUDFCreation {
> ```

nit: DO block seems orthogonal to UDF creation, so I don't think we should be checking it. We might want to provide a separate "disable DO block" sqlsmith option though, but I'd hold off on that until we find a test that needs it.


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/151983)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @DrewKimball and @ZhouXing19)

___
*[`pkg/internal/sqlsmith/plpgsql.go` line 144 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/151983#-OXyVm3Q6cs9vngiks6I:-OXyYc3sCvMaFFmVBvvV:b5swr7e) ([raw file](https://github.com/cockroachdb/cockroach/blob/7be608d016075aa34516ed500046491d605ee922/pkg/internal/sqlsmith/plpgsql.go#L144)):*
<details><summary><i>Previously, ZhouXing19 (Jane Xing) wrote…</i></summary><blockquote>

> Given that DO block is only allowed as top-level SQL stmt
hmm `DO` can actually be part of the function body, see [example](https://github.com/cockroachdb/cockroach/blob/1f851288a24751906480655880cb60a2f9d1b2f2/pkg/sql/logictest/testdata/logic_test/do#L101-L112). But not sure if i misunderstood what you mean by `top-level SQL stmt`.
</blockquote></details>

Oh, I was looking at `tree.DoBlock` from `tree/do.go` where I saw this comment. Make sense that it's also allowed to be within body of a routine, and we should try to generate both variants. (A couple of recent panics where the result of the DO block that wasn't inside a routine itself.)


<!-- Sent from Reviewable.io -->

ZhouXing19 committed:
sqlsmith: support DO block for plpgsql

A `DO` block can be either a top-level sql statement or be part of the
plpgsql function body. In this commit we introduce the DO block for
both of these 2 usages. We also introduced `DisableDoBlocks` which
disable both usages of a DO block.

Release note: None

yuzefovich reviewed (approved):
Nice! <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 4 of 4 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/151983#-:-OXzmhbu6xm8af_RXnQF:be8rotp)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @DrewKimball and @ZhouXing19)

___
*[`pkg/internal/sqlsmith/plpgsql.go` line 144 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/151983#-OXyVm3Q6cs9vngiks6I:-OXzlOLR0tgVZ2Ua8_is:bygpzeo) ([raw file](https://github.com/cockroachdb/cockroach/blob/7be608d016075aa34516ed500046491d605ee922/pkg/internal/sqlsmith/plpgsql.go#L144)):*
<details><summary><i>Previously, ZhouXing19 (Jane Xing) wrote…</i></summary><blockquote>

> Perhaps we should consider two variants: `makeDoBlockNonMutating` (which disallows mutations) and a general `makeDoBlock` (which allows any statements).

However, looking more closely at how `makePLpgSQLBlock()` works—it calls `makePLpgSQLExecSQL()`, which in turn uses `Smither.makeSQLStmtForRoutine()` to generate the statements inside the DO block—it seems that we already enforce the `Smither.disableMutations` check there. In other words, a DO block automatically adapts its contents based on `disableMutations`.

Given that, I’m not sure it’s necessary to split into `makeDoBlockNonMutating` and `makeDoBlock`. Would it be acceptable to simply include `makeDoBlock` in both `mutatingStatements` and `nonMutatingStatements`? Or perhaps it deserves its own category (similar to how `makeCreateFunc` doesn’t fall neatly into either group)?
</blockquote></details>

Yeah, good point, I agree that existing `disableMutations` should work.

Including `makeDoBlock` into both mutating and non-mutating stmts sounds good to me - I'd only consider reducing the weights of the Do block, maybe to 1 in both cases, WDYT? The larger the weight, the more likely the stmt will be chosen, and we probably want Do blocks to be generated at roughly the same frequency as other "unusual" stmts like CREATE STATS. The fact that we're including Do blocks into both mutating and non-mutating sets will make Do blocks more frequent than "pure weight 1" stmts.
___
*[`pkg/internal/sqlsmith/plpgsql.go` line 413 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/151983#-OXzm3hX8vUPSTLYpWg0:-OXzm3hX8vUPSTLYpWg1:bp5pioz) ([raw file](https://github.com/cockroachdb/cockroach/blob/56cc6529d063fc1433a722498ba3dc9626b7dfea/pkg/internal/sqlsmith/plpgsql.go#L413)):*
> ```Go
> 		return nil, false
> 	}
> 	block, _ := makePLpgSQLBlock(s, scope)
> ```

nit: you could just use `s.makePLpgSQLBlock(scope)` to avoid having unnecessary `ok` return parameter.


<!-- Sent from Reviewable.io -->

ZhouXing19 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/151983)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @DrewKimball and @yuzefovich)

___
*[`pkg/internal/sqlsmith/plpgsql.go` line 144 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/151983#-OXyVm3Q6cs9vngiks6I:-OY1NqO38wh5gRKg8cVm:bydhcq1) ([raw file](https://github.com/cockroachdb/cockroach/blob/7be608d016075aa34516ed500046491d605ee922/pkg/internal/sqlsmith/plpgsql.go#L144)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

Yeah, good point, I agree that existing `disableMutations` should work.

Including `makeDoBlock` into both mutating and non-mutating stmts sounds good to me - I'd only consider reducing the weights of the Do block, maybe to 1 in both cases, WDYT? The larger the weight, the more likely the stmt will be chosen, and we probably want Do blocks to be generated at roughly the same frequency as other "unusual" stmts like CREATE STATS. The fact that we're including Do blocks into both mutating and non-mutating sets will make Do blocks more frequent than "pure weight 1" stmts.
</blockquote></details>

Good idea, reduced both weights to 1.


<!-- Sent from Reviewable.io -->

DrewKimball reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Nice work on this!

> @DrewKimball reviewed 2 of 4 files at r2, 2 of 2 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/151983#-:-OYhaSlA9hCN5BkNUQak:b-wrpl1f)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @yuzefovich)



<!-- Sent from Reviewable.io -->

yuzefovich commented:
Nice! Looks like the failure in CI already exposes #151848. Let me cherry-pick this commit and stress my fix for that issue in #151849.

yuzefovich commented:
#151849 and #152201 have been merged, so this change should be good to go. It might be a good idea to rebase the PR on top of latest master to trigger an extra run of CI before merging.
